apps/src/studio/studio.js

```diff
@@ -4307,10 +4329,51 @@ Studio.allGoalsVisited = function() {
   return finishedGoals === Studio.spriteGoals_.length;
 };
 
+Studio.checkRequiredForSuccess = function() {
+
+  if (! level.requiredForSuccess) {
+    return { achieved: false, message: null };
+  }
+
+  var required = level.requiredForSuccess;
+
+  if (required.setSprite && !Studio.hasSetSprite) {
+    return { achieved: false, message: studioMsg.failedHasSetSprite() };
+  }
+
+  if (required.setBotSpeed && !Studio.hasSetBotSpeed) {
+    return { achieved: false, message: studioMsg.failedHasSetBotSpeed() };
+  }
+
+  if (required.touchAllItems && Studio.items.length > 0) {
+    return { achieved: false, message: studioMsg.failedTouchAllItems() };
+  }
+
+  if (required.scoreMinimum && Studio.playerScore < required.scoreMinimum) {
+    return { achieved: false, message: studioMsg.failedScoreMinimum() };
+  }
+
+  if (required.removedItemCount && Studio.removedItemCount < required.removedItemCount) {
+    return { achieved: false, message: studioMsg.failedRemovedItemCount() };
+  }
+
+  if (required.setActivity &&
+      !(Studio.setActivityRecord && 
+        Studio.setActivityRecord[required.setActivity.itemType] &&
+        Studio.setActivityRecord[required.setActivity.itemType][required.setActivity.activityType])) {
+    return { achieved: false, message: studioMsg.failedSetActivity() };
+  }
+
+  return { achieved: true, message: null };
+};
+
+
 var checkFinished = function () {
 
   var hasGoals = Studio.spriteGoals_.length !== 0;
   var achievedGoals = Studio.allGoalsVisited();
+  var hasRequiredForSuccess = level.requiredForSuccess ? true : false;
+  var requiredForSuccess = Studio.checkRequiredForSuccess();
```

Comment:
Seems like it would be nice if we can have hasRequiredForSuccess be part of the object returned by checkRequiredForSuccess. Maybe as simple as adding a boolean field calls "exists"?